plugin/src/main/java/org/opensearch/ml/processor/MLInferenceIngestProcessor.java

```diff
@@ -0,0 +1,368 @@
+    public static final String DEFAULT_OUTPUT_FIELD_NAME = "inference_results";
+
+    protected MLInferenceIngestProcessor(
+        String modelId,
+        List<Map<String, String>> inputMaps,
+        List<Map<String, String>> outputMaps,
+        Map<String, String> modelConfig,
+        String tag,
+        String description,
+        boolean ignoreMissing,
+        ScriptService scriptService,
+        Client client
+    ) {
+        super(tag, description);
+        this.inferenceProcessorAttributes = new InferenceProcessorAttributes(modelId, inputMaps, outputMaps, modelConfig);
+        this.ignoreMissing = ignoreMissing;
+        this.scriptService = scriptService;
+        this.client = client;
+    }
+
+    /**
+     * overwrite in this execute method,
+     * when batch inference is available,
+     * to support async multiple predictions.
+     */
+    @Override
+    public void execute(IngestDocument ingestDocument, BiConsumer<IngestDocument, Exception> handler) {
+
+        List<Map<String, String>> processInputMap = inferenceProcessorAttributes.getInputMaps();
+        List<Map<String, String>> processOutputMap = inferenceProcessorAttributes.getOutputMaps();
+        int round = (processInputMap != null) ? processInputMap.size() : 0;
+
+        process_predictions(ingestDocument, handler, processInputMap, processOutputMap, 0, round);
+
+    }
+
+    @Override
+    public IngestDocument execute(IngestDocument ingestDocument) throws Exception {
+        return ingestDocument;
+    }
+
+    /**
+     * process predictions for one model for multiple rounds of predictions
+     * ingest documents after prediction rounds are completed,
+     * when no input mappings provided, default to add all fields to model input fields,
+     * when no output mapping provided, default to output as
+     * "inference_results" field (the same format as predict API)
+     */
+    private void process_predictions(
+        IngestDocument ingestDocument,
+        BiConsumer<IngestDocument, Exception> handler,
+        List<Map<String, String>> processInputMap,
+        List<Map<String, String>> processOutputMap,
+        int i,
+        int round
+    ) {
+        if (i >= round && i != 0) {
+            handler.accept(ingestDocument, null);
+        } else {
+
+            Map<String, String> modelParameters = new HashMap<>();
+
+            if (inferenceProcessorAttributes.getModelConfig() != null) {
+                modelParameters.putAll(inferenceProcessorAttributes.getModelConfig());
+            }
+            // when no input mapping is provided, default to read all fields from documents as model input
+            if (round == 0) {
+                Set<String> documentFields = ingestDocument.getSourceAndMetadata().keySet();
+                for (String field : documentFields) {
+                    getMappedModelInputFromDocuments(ingestDocument, modelParameters, field, field);
+                }
+
```

Comment:
This part should be moved from this recursive function, we don't have to execute it in each iteration.

Reply:
In each round of the prediction, it's better to initiate a new map of modelParameters, to pass to the prediction task, to avoid the modelParameters get overrides in different rounds.